Add typless endpoints for get_source and exist_source

cbuescher opened the pull request:
This PR is still work in progress since I expect several test failures but I want to get it into CI.

Commit:
WIP

Commit:
iter

elasticmachine commented:
Pinging @elastic/es-search

jtibshirani commented on server/src/test/java/org/elasticsearch/rest/action/document/RestGetSourceActionTests.java:
```diff
@@ -21,26 +21,51 @@
 import static org.hamcrest.Matchers.equalTo;
+import static org.mockito.Mockito.mock;
 
 public class RestGetSourceActionTests extends ESTestCase {
```
There is now a base class `RestActionTestCase` that helps remove some of the test boilerplate.

jtibshirani commented on server/src/test/java/org/elasticsearch/rest/action/document/RestGetSourceActionTests.java:
```diff
@@ -49,6 +74,39 @@ public static void cleanupReferences() {
+        Map<String, String> params = new HashMap<>();
+        params.put("type", "some_type");
+        RestRequest request = new FakeRestRequest.Builder(xContentRegistry())
+            .withMethod(randomFrom(Arrays.asList(Method.GET, Method.HEAD)))
```
Since these tests are so fast + simple, maybe we could just test both methods every time. I don't think this will be much extra code, as you could re-use the same `FakeRestRequest.Builder`.

jtibshirani commented on server/src/main/java/org/elasticsearch/action/get/GetRequest.java:
```diff
@@ -30,14 +30,15 @@
 import static org.elasticsearch.action.ValidateActions.addValidationError;
 
 /**
- * A request to get a document (its source) from an index based on its type (optional) and id. Best created using
```
I recently merged a PR to deprecate types in `get` requests, so hopefully these changes will already be covered.

jtibshirani commented on docs/reference/docs/get.asciidoc:
```diff
@@ -42,7 +42,7 @@ The API also allows to check for the existence of a document using
 
 [source,js]
 --------------------------------------------------
-HEAD twitter/_doc/0
```
The document CRUD APIs are retaining the `_doc` path component (so this actually shouldn't change).

jtibshirani commented on rest-api-spec/src/main/resources/rest-api-spec/api/exists_source.json:
```diff
@@ -3,8 +3,8 @@
     "url": {
-      "path": "/{index}/{type}/{id}/_source",
-      "paths": ["/{index}/{type}/{id}/_source"],
+      "path": "/{index}/{id}/_source",
```
In our proposal, we actually decided the new endpoint should be `/{index}/_source/{id}`. Here was our reasoning for not simply removing the type component (as in `{index}/{id}/_source`):
- Unlike types, IDs are allowed to contain underscores, and there could be collisions with other endpoints like `{index}/_doc/{id}`.
- It doesn’t work cleanly for all URLs, for example 'termvectors' where it introduces collisions between the different endpoints.

jtibshirani commented on rest-api-spec/src/main/resources/rest-api-spec/test/get_source/40_routing.yml:
```diff
@@ -26,7 +26,6 @@
  - do:
       get_source:
           index:   test_1
-          type:    test
```
I think that in order to complete successfully, we'll need to have a `skip` section that limits these tests to 7.0.

We'll also want to retain the old typed versions of these tests, as we've done for the other deprecated APIs. For example, we can introduce a new file `41_routing_with_types` to continue to test the typed calls.

jtibshirani reviewed (commented):
I see this is a WIP, just wanted to give some early comments to help clarify a few points.

Commit:
Merge branch 'master' into types-removal-get_source

Commit:
iter

Commit:
add type and typless yml tests

Commit:
Merge branch 'master' into types-removal-get_source

Commit:
iter

Commit:
Add change in enpoint mapping for hlrc

Commit:
Fix Netty4HeadBodyIsEmptyIT

cbuescher commented:
@elasticmachine run gradle build test 1

cbuescher commented:
@elasticmachine run gradle build test 1

cbuescher commented:
@elasticmachine run gradle build 1

cbuescher commented:
@elasticmachine run gradle build test 1

cbuescher commented:
run the gradle build tests 1

jtibshirani commented on server/src/main/java/org/elasticsearch/rest/action/document/RestGetSourceAction.java:
```diff
@@ -62,7 +70,13 @@ public String getName() {
-        final GetRequest getRequest = new GetRequest(request.param("index"), request.param("type"), request.param("id"));
+        final GetRequest getRequest;
+        if (request.hasParam("type")) {
+            deprecationLogger.deprecatedAndMaybeLog("get_source_type", TYPES_DEPRECATION_MESSAGE);
```
Really small comment, but for consistency this would be `get_source_with_types`.

cbuescher replied:
I'm suprised this gets logges somewhere, so far I thought this only needs to be a unique key. But I'll change.

jtibshirani commented on server/src/main/java/org/elasticsearch/action/get/GetRequest.java:
```diff
@@ -109,9 +108,6 @@ public GetRequest(String index, String id) {
     @Override
     public ActionRequestValidationException validate() {
         ActionRequestValidationException validationException = super.validateNonNullIndex();
-        if (Strings.isEmpty(type)) {
```
Do we need this change? I think the type should always be present, and a 'typeless' request is represented by using `_doc`.

cbuescher replied:
I guess we can change it later when we remove "_doc" as well. Thats the plan, isn't it?

jtibshirani replied:
Right, we can eventually update these requests to not contain the `type` field at all.

jtibshirani commented on docs/reference/docs/get.asciidoc:
```diff
@@ -1,9 +1,9 @@
 == Get API
 
-The get API allows to get a typed JSON document from the index based on
+The get API allows to get a JSON document from the index based on
```
Thanks for cleaning this up!

jtibshirani commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java:
```diff
@@ -267,7 +267,12 @@ private static Request getStyleRequest(String method, GetRequest getRequest) {
 
     static Request sourceExists(GetRequest getRequest) {
-        Request request = new Request(HttpHead.METHOD_NAME, endpoint(getRequest.index(), getRequest.type(), getRequest.id(), "_source"));
+        String endpoint = new EndpointBuilder()
```
If the request has a type, we should make sure to use the old endpoint (see `RequestConverters#explain` for example). We should also update `RequestConvertersTests` to check both the typeless + typed versions (as in `testExplain` and `testExplainWithType`).

cbuescher replied:
Is that for mixed cluster scenarios? Just so I understand it better for next time.

cbuescher replied:
As we said above, GetRequest has always a type, if it isn't provided it is "_doc". Would we always want to use the old deprecated endpoint then?

jtibshirani replied:
> Is that for mixed cluster scenarios?

Right, this allows us to support a scenario where some nodes are running 6.x, and don't understand the new typeless APIs. Also, our goal is to not make breaking changes to the HLRC behavior in 7.0 -- calling the old typed endpoint when a type is provided in a request allows us to maintain the exact same behavior as in 6.x, where we require a match on the type name.

> Would we always want to use the old deprecated endpoint then?

If HLRC requests are constructed without types (as in `new GetRequest("index", "id")`), then we should use the typeless endpoints so that users can move to the new APIs + avoid deprecation warnings. I think of it as an internal implementation detail that a typeless get request executes by pretending the type is `_doc`. Note that if we had two separate `GetRequest` classes (one for the HLRC, one for the server), then we probably would have implemented this differently. For example in the HLRC `TermVectorsRequest`, we chose to use 'null' for the type in typeless requests, since this is a separate class from the server version.

I hope this makes sense, please feel free to ping me and I can try to explain further.

jtibshirani reviewed (commented):
A few more comments, hope I'm not jumping in too early with reviews!

Commit:
Merge branch 'master' of github.com:elastic/elasticsearch into types-removal-get_source

Commit:
Addressing review comments

cbuescher commented:
@jtibshirani thanks for the review, I removed the WIP label since the last round of CI tests passed. I added the requested changes, can you please take another look if this looks good to you now?

cbuescher commented:
run the gradle build tests 2

jtibshirani commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java:
```diff
@@ -156,6 +157,54 @@ public void testGetWithType() {
+        String id = randomAlphaOfLengthBetween(3, 10);
+        final GetRequest getRequest;
+        if (randomBoolean()) {
+            getRequest = new GetRequest(index, id);
```
Instead of using randomization, we've been creating two separate tests to make sure that we always test both typed + typeless requests (see `testExplain`/ `testExplainWithTypes` for example). This is something we settled on recently, my reference PR was probably out-of-date and led you down this path.

cbuescher replied:
Actually it was a consciuos decision because I don't think the double testing is really necessary, we run so many tests in CI that this will trigger both variants quickly. But I will add a second test if that is what you decided.

jtibshirani replied:
Got it. I agree we should go with the two tests now for consistency.

jtibshirani commented on server/src/test/java/org/elasticsearch/action/get/GetRequestTests.java:
```diff
@@ -40,7 +40,7 @@ public void testValidation() {
 
             assertThat(validate, not(nullValue()));
-            assertThat(validate.validationErrors(), hasItems("type is missing", "id is missing"));
+            assertThat(validate.validationErrors(), hasItems("id is missing"));
```
I don't think we need this test change.

cbuescher replied:
Yes, thats a left-over from the validation I removed and then added again. Will update.

jtibshirani commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java:
```diff
@@ -268,8 +268,14 @@ private static Request getStyleRequest(String method, GetRequest getRequest) {
-
+        String optionalType = getRequest.type();
+        String endpoint;
+        if (optionalType == null || optionalType.equals(MapperService.SINGLE_MAPPING_NAME)) {
```
Two very small comments: the indentation is a bit off, and I don't think we need to check if the type is null because `GetRequest#validate` ensures that it isn't.

jtibshirani reviewed (commented):
A few minor comments, it looks good to merge once these are addressed.

Commit:
another iteration

cbuescher commented:
@jtibshirani pushed another commit adressing the last review so you can take another look if you have time.

Commit:
Merge branch 'master' into types-removal-get_source
